// coding=utf-8
// Copyright 2022 Foivos Tsimpourlas and Chris Cummins.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto2";

package benchpress;

option go_package = "benchpresspb";
option java_multiple_files = true;
option java_outer_classname = "SamplerProto";
option java_package = "com.benchpress";

import "deeplearning/benchpress/proto/model.proto";
import "deeplearning/benchpress/proto/corpus.proto";
import "deeplearning/benchpress/proto/active_learning.proto";

// The specification of a benchpress sampler.
message Sampler {
  // The initial text to the seed the language model with. Each sample will
  // begin with this text.
  oneof sample_feed {
    // Simple string
    string        start_text      = 1;
    // Sample from training set
    bool          train_set       = 2;
    // Sample from validation set
    bool          validation_set  = 3;
    // Create set with new specs from original corpus
    ValidationSet sample_set      = 4;
    // Specify a whole new corpus to encode and sample with new specs.
    SampleCorpus  sample_corpus   = 5;
    // Sample live by getting input() from user.
    bool          live_sampling   = 6;
  }
  // The sampling batch size.
  // TODO(cec): Always sample with max batch size.
  optional int32 batch_size = 7;
  // The length of sampling sequences.
  optional int32 sequence_length = 8;
  // The sampling temperature. Must be >= 0. A recommended starting value is
  // 1000000 (i.e. 1.0 in real values).
  optional int32 temperature_micros = 9;
  // The criteria that determine when to terminate a sample, in the order in
  // which they will be executed. Duplicates are allowed, for example if you
  // would like to have symmetrical token depth counters for two pairs of
  // tokens.
  repeated SampleTerminationCriterion termination_criteria = 10;
}

// Specification of a new sample corpus to get feeds from.
message SampleCorpus {
  optional CorpusConfig corpus_config = 1;
  oneof input_corpus {
    Corpus corpus     = 2;
    string start_text = 3;
  }
}

message CorpusConfig {
  // Encoding and data masking configuration for sample corpus.
  // sampling type can be 'normal', 'online', 'active'
  // normal: corpus is pre-masked, then fed for predictions.
  // online: A datapoint is requested from original corpus and masked on the fly.
  // active: Same as online. Active learning is applied between sample and target features
  oneof sampling_type {
    bool           normal = 1;
    bool           online = 2;
    ActiveSampler  active = 3;
  }
  optional int32  max_predictions_per_seq = 4;
  optional float  masked_lm_prob          = 5;
  oneof mask_technique {
    Mask mask = 6;
    Hole hole = 7;
    MaskSeq mask_seq = 8;
  }
}

// Active Sampling specifications message.
message ActiveSampler {
  optional int32  active_search_depth   = 1;
  optional int32  active_search_width   = 2;
  optional float  active_dropout_prob   = 3;
  optional int32  batch_size_per_feed   = 4;
  optional string feature_space         = 5;
  oneof target_features {
    string target                = 6;
    ActiveLearner active_learner = 7;
  }
}

// Criteria used for determining when to stop sampling.
message SampleTerminationCriterion {
  oneof criterion {
    MaxTokenLength maxlen = 1;
    SymmetricalTokenDepth symtok = 2;
  }
}

message MaxTokenLength {
  // The maximum length of a sample, as a number of tokens. The length of the
  // Sampler.start_text counts towards this total.
  optional int32 maximum_tokens_in_sample = 1;
}

message SymmetricalTokenDepth {
  optional string depth_increase_token = 1;
  optional string depth_decrease_token = 2;
}
